### DISCUSSION TITLE
Add more comprehensive examples of best practise eslint.config.js's for v9 migration

### DISCUSSION THREAD

**Discussion by TeemuKoivisto** [eslint/eslint#18377] [Ideas]
So I just updated my eslint to v9 and with it the default configuration file was changed to `eslint.config.js`

As someone who hasn't known about the configuration change at all, it was quite painful to realize how big of a change this major version update became. It wasn't immediately obvious where to find the new link for migrating https://eslint.org/docs/latest/use/configure/migration-guide and even from there reading such a huge article started to feel quite tedious as this was not what I wanted in the first place. 

Some changes were easy like moving the parser and env blocks (although I'm not sure does env need to be renamed or nested under globals). However, with `extends` and `plugins` blocks I had no idea is importing them necessary anymore or what's the recommend best practise. It seems `eslint:recommended` was moved to` "@eslint/js".configs.recommended` so this added to the confusion. I am using TypeScript with Svelte and from quick searching I couldn't find an example I could have examined.

Here's a part of my config to show what I'm talking about:

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["svelte3", "@typescript-eslint", "eslint-plugin-import", "eslint-plugin-prettier"],
  "extends": ["eslint:recommended", "plugin:@typescript-eslint/recommended", "prettier"],
  "env": {
    "es6": true,
    "browser": true,
    "node": true
  },
  "parserOptions": {
    "sourceType": "module"
  },
  "overrides": [
    {
      "files": ["*.svelte"],
      "processor": "svelte3/svelte3"
    }
  ],
  "settings": {
    "svelte3/typescript": true
  },
  "rules": ...
}
```

As it started to seem I was wasting my time when I could be productive, I reverted back to v8 and kept my old config since linting is  quite minor problem in my case especially as I use prettier.

But as this migration will become unavoidable in the future, I'd really wish for more comprehensive `eslint.config.js` examples directly in the documentation to expedite the process. It's a fact most developers don't have the time nor motivation to read long migration docs and would rather just copy paste the new best practise example and figure out from there how to tweak it. Listing just a few at the top of the migration page would probably be enough; TS, prettier and some frontend framework related.

**Comment by MAST1999** [comment]
I would agree that the migration guide needs to expand on how to migrate the extends and plugin blocks.

**Comment by nzakas** [comment]
Thanks for the feedback. It looks like there are several examples using `plugins`, starting from here:
https://eslint.org/docs/latest/use/configure/migration-guide#importing-plugins-and-custom-parsers

We do also have a section on `extends` here:
https://eslint.org/docs/latest/use/configure/migration-guide#predefined-and-shareable-configs

It seems like these cover most cases. 

Because the ESLint ecosystem is so vast, it's difficult to know what kind of examples will work for everyone. In general, we don't recommend using `eslint-plugin-prettier` or `eslint-config-prettier`, so probably not the best example to highlight.

We are working on a migration tool that should help with some of this.

In the meantime, if you think there are gaps in those two sections of docs, let me know what you think should be included.

**Reply by symplytheo** [reply]
Why did you say you don't recommend using eslint-plugin-prettier and eslint-config-prettier, especially as it's something we have been using with the previous versions?

**Reply by nzakas** [reply]
We've always recommended people avoid those, that's not new. Because they run Prettier inside of ESLint it is significantly slower and also can cause conflicts with ESLint rules. It's much better to run Prettier as a separate task after ESLint is run.

**Reply by TeemuKoivisto** [reply]
I see, thanks. Yes I think the major issue is the lack of proper migration path for the silly downstream user. I apologize for the rather long reply but here we go.

So this is my previous `.eslintrc.json`: https://gist.github.com/TeemuKoivisto/87c90bcdca5d6736fb99bf53b262457e

This is the new `eslint.config.js` I picked from create SvelteKit app: https://gist.github.com/TeemuKoivisto/05c7eee930ecb468327debd5124dcba3

You can see the rather drastic changes. Certainly, some of the plugins were outdated but still I could not have figured this out in reasonable time by myself. Thank god for the boilerplate authors.

However, even with this working config I find it immensely annoying that my old command to run eslint: `eslint --ignore-path .gitignore --fix --ext .js,.cjs,.ts,.tsx,.svelte,.css,.scss,.html .` no longer works and I'm thrown into https://eslint.org/docs/latest/use/command-line-interface page with no actual explanation what I should use instead. That whole page is 30 viewports tall so you can imagine I feel quite overwhelmed from the massive information overload.

I really want to ignore my `.gitignore` paths but I don't understand should I now import it manually inside the `eslint.config.js` and pass it as string to `ignores` array?

I'm sure it has been a huge hassle to make v9 and all the documentation involved and thanks for the great work. It's just I feel the complexity is leaking a little bit too much for the common folk who have little context what's going on.

**Reply by nzakas** [reply]
> However, even with this working config I find it immensely annoying that my old command to run eslint: `eslint --ignore-path .gitignore --fix --ext .js,.cjs,.ts,.tsx,.svelte,.css,.scss,.html .` no longer works and I'm thrown into https://eslint.org/docs/latest/use/command-line-interface page with no actual explanation what I should use instead. That whole page is 30 viewports tall so you can imagine I feel quite overwhelmed from the massive information overload.

Here's the docs on migrating `--ext`:
https://eslint.org/docs/latest/use/configure/migration-guide#--ext

Yes, to use the `.gitignore` file, you now need to include it in your `eslint.config.js` file. It's a fair point that this probably reasonably common so we should be able to make a utility to do that more easily. I'll look into it.

**Reply by nzakas** [reply]
https://github.com/eslint/rewrite/issues/44

**Reply by TeemuKoivisto** [reply]
Oh wow, thanks! Much appreciated

**Reply by joelspadin** [reply]
> We've always recommended people avoid those, that's not new. Because they run Prettier inside of ESLint it is significantly slower and also can cause conflicts with ESLint rules. 

That makes sense for eslint-plugin-prettier, but eslint-config-prettier is just a list of rules to turn off to make sure running prettier doesn't make changes that would fail eslint checks. As far as I can tell, there's nothing in it that actually runs prettier.
